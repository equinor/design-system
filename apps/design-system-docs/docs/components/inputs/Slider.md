# Slider

When expanded use tab to review current page headings and press enter or space to navigate to the selected section

INPUTS

# Slider

A _slider_ is a flexible input that allows users to adjust predefined values.

  

**More examples available in** **.**

## Guidelines

Sliders reflect values along a bar where users can select a single value or a range of values. Sliders reflect the current state of the values they control, and value changes are immediate to the content in which they belong.



# Design

When expanded use tab to review current page headings and press enter or space to navigate to the selected section

INPUTS

# Slider

A _slider_ is a flexible input that allows users to adjust predefined values.

## Implementation in Figma

### Instructions

1.  In Figma go to the **Assets Panel** and search for **slider**.
    
2.  Drag and drop the component in your frame.
    
3.  Rename and resize the component if needed.
    
4.  Choose the variant from the **Design Panel**.



# Code

When expanded use tab to review current page headings and press enter or space to navigate to the selected section

INPUTS

# Slider

A _slider_ is a flexible input that allows users to adjust predefined values.