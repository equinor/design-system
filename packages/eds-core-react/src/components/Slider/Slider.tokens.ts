import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  colors: {
    ui: {
      background__light: { rgba: backgroundColor },
      background__default: { rgba: backgroundColorDefault },
      background__medium: { rgba: backgroundColorMedium },
    },
    interactive: {
      primary__resting: { rgba: indicatorColor },
      primary__hover_alt: { rgba: primaryHoverAlt },
      primary__hover: { rgba: primaryHover },
      focus: { rgba: focusOutlineColor },
      disabled__fill: { rgba: backgroundColorDisabled },
      disabled__border: { rgba: backgroundBorderDisabled },
    },
    text: {
      static_icons__tertiary: { rgba: textColor },
    },
  },
  typography: { paragraph },
  interactions: {
    focused: { width: focusOutlineWidth },
  },
} = tokens

type SliderToken = ComponentToken

export const slider: SliderToken = {
  background: backgroundColorDefault,
  entities: {
    track: {
      background: backgroundColor,
      height: '4px',
      spacings: {
        top: '24px',
        bottom: '9px',
      },
      entities: {
        indicator: {
          background: indicatorColor,
          states: {
            hover: {
              background: primaryHover,
            },
            disabled: {
              background: backgroundBorderDisabled,
            },
          },
        },
      },
      states: {
        hover: {
          background: backgroundColorMedium,
        },
      },
    },
    output: {
      typography: {
        ...paragraph.caption,
        color: textColor,
      },
    },
    handle: {
      background: backgroundColorDefault,
      height: '12px',
      width: '12px',
      border: {
        type: 'border',
        color: indicatorColor,
        radius: '50%',
        width: '2px',
        style: 'solid',
      },
      states: {
        focus: {
          outline: {
            type: 'outline',
            color: focusOutlineColor,
            width: focusOutlineWidth,
            style: 'dashed',
            offset: '3px',
          },
        },
        hover: {
          background: primaryHoverAlt,
          border: {
            type: 'border',
            color: primaryHover,
          },
        },
      },
    },
    dot: {
      height: '6px',
      width: '6px',
      border: {
        type: 'border',
        color: backgroundColorMedium,
        width: '1px',
        style: 'solid',
        radius: '50%',
      },
      spacings: {
        bottom: '8px',
      },
    },
  },
  states: {
    disabled: {
      background: backgroundColorDisabled,
      border: {
        type: 'border',
        color: backgroundColorMedium,
      },
    },
  },
}
