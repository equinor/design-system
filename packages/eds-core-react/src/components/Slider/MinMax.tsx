import { forwardRef, HTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '../../utils'
import { slider as tokens } from './Slider.tokens'

const {
  entities: { track, output },
} = tokens

const StyledMinMax = styled.span`
  grid-row: 2;
  ${typographyTemplate(output.typography)}
  position: absolute;
  left: 2px;
  top: ${track.spacings.top};
  /* Avoid track hover on minmax mouse over */
  pointer-events: none;
  text-align: left;
  /** Center align the text with the dot */
  transform: translate(calc(-1 * calc((100% - 8px) / 2)));
  &:last-child {
    left: auto;
    right: 2px;
    transform: translate(calc((100% - 8px) / 2));
  }
`

type MinMaxProps = {
  /** Children is required */
  children: ReactNode
} & HTMLAttributes<HTMLSpanElement>

export const MinMax = forwardRef<HTMLSpanElement, MinMaxProps>(function MinMax(
  { children },
  ref,
) {
  return <StyledMinMax ref={ref}>{children}</StyledMinMax>
})
