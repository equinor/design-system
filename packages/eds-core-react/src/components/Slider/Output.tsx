import {
  forwardRef,
  OutputHTMLAttributes,
  ReactNode,
  CSSProperties,
} from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '@equinor/eds-utils'
import { slider as tokens } from './Slider.tokens'

const {
  entities: { output },
} = tokens

const StyledOutput = styled.output`
  --realWidth: calc(100% - 12px);
  --background: rgb(0 0 0 / 0.8);
  width: fit-content;
  position: absolute;
  display: flex;
  align-items: center;
  border-radius: 4px;
  z-index: 1;
  ${typographyTemplate(output.typography)};
  color: white;
  background: var(--background);
  padding: 4px 8px;
  bottom: calc(100% + 8px);
  pointer-events: none;
  /* Calculate the distance on the track*/
  margin-left: calc((var(--val) - var(--min)) / var(--dif) * var(--realWidth));
  /* Idea: Transform negative ((width of outline elem - handle width) / 2 (half of width for centering)) */
  transform: translate(calc(-1 * calc(var(--realWidth) / 2)));
  grid-row: 2;
  grid-column: 1 / -1;
  opacity: var(--showTooltip);
`

const TooltipArrow = styled.svg`
  width: 6px;
  height: 8px;
  position: absolute;
  fill: var(--background);
  top: calc(100% - 1px);
  rotate: -90deg;
  translate: -50%;
  left: 50%;
`

type OutputProps = {
  /** Value */
  value: number
  /** HtmlFor */
  htmlFor: string
  /** Children are required */
  children: ReactNode
} & OutputHTMLAttributes<HTMLOutputElement>

export const Output = forwardRef<HTMLOutputElement, OutputProps>(
  function Output({ children, value, htmlFor }, ref) {
    return (
      <StyledOutput
        ref={ref}
        style={{ '--val': value } as CSSProperties}
        htmlFor={htmlFor}
      >
        {children}
        <TooltipArrow>
          <path d="M0.504838 4.86885C-0.168399 4.48524 -0.168399 3.51476 0.504838 3.13115L6 8.59227e-08L6 8L0.504838 4.86885Z" />
        </TooltipArrow>
      </StyledOutput>
    )
  },
)
