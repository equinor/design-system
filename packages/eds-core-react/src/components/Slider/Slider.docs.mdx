import { Links } from './../../../.storybook/components'
import { Primary, Canvas, Controls } from '@storybook/blocks'
import * as ComponentStories from './Slider.stories'

# Slider

Is a flexible input that allows users to adjust predefined values.

<Links 
    ariaUrl="https://www.w3.org/WAI/ARIA/apg/patterns/slider/"
    figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17490%3A95"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/550c4f-slider/b/86826a"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Slider/Slider.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Primary />
<Controls />

## Usage

Sliders reflect the current state of the settings they control.

<ul>
    <li>Sliders should be used for making selections from a range of values.</li>
    <li>When interacting with a slider, changes should be reflected back to a user immediately.</li>
    <li>Sliders should present the full range of choices that are available to a user.</li>
</ul>

```tsx
import { Slider } from '@equinor/eds-core-react'

<Slider value={4} min={0} max={10} aria-label="simple-slider" />
```

## Examples

### Simple slider

<Canvas of={ComponentStories.SimpleSlider} />

### Label always visible

<Canvas of={ComponentStories.LabelAlwaysOn} />

### Simple slider with steps

<Canvas of={ComponentStories.SimpleSliderWithSteps} />

### Range slider

<Canvas of={ComponentStories.RangeSlider} />

### Range slider with committed step

<Canvas of={ComponentStories.RangeSliderWithCommittedStep} />

### Range slider with interval

<Canvas of={ComponentStories.RangeSliderWithInterval} />

### Range slider with dates

<Canvas of={ComponentStories.RangeSliderWithDates} />

### Disabled

<Canvas of={ComponentStories.Disabled} />

### Hide active track
`hideActiveTrack` visually removes the moss green fill color from the "active" part of the track
<Canvas of={ComponentStories.HideActiveTrack} />
