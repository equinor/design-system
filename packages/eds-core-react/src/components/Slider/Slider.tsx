import {
  forwardRef,
  useState,
  useRef,
  HTMLAttributes,
  MouseEvent,
  TouchEvent,
  KeyboardEvent,
  useEffect,
  ChangeEvent,
  useCallback,
} from 'react'
import styled, { css } from 'styled-components'
import { slider as tokens } from './Slider.tokens'
import { MinMax } from './MinMax'
import { Output } from './Output'
import { SliderInput } from './SliderInput'
import { bordersTemplate, useId } from '@equinor/eds-utils'

const {
  entities: { track, handle, dot },
} = tokens

const fakeTrackBg = css`
  background-image: url("data:image/svg+xml,<svg xmlns='http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'><rect x='0' y='11' fill='${track.background}' width='100%' height='4' rx='2' /></svg>");
  background-size: cover;
  background-repeat: no-repeat;
`

const fakeTrackBgHover = css({
  backgroundImage: `url("data:image/svg+xml,<svg xmlns='http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'><rect x='0' y='11' fill='${track.states.hover.background}' width='100%' height='4' rx='2' /></svg>")`,
})

const trackFill = css`
  grid-column: 1 / span 2;
  grid-row: 2;
  height: ${track.height};
  margin-bottom: ${track.spacings.bottom};
  align-self: end;
  content: '';
`

const wrapperGrid = css`
  display: grid;
  grid-template-rows: max-content 24px;
  grid-template-columns: 1fr 1fr;
  width: 100%;
  position: relative;
`
type RangeWrapperProps = {
  $valA: number
  $valB: number
  $min: number
  $max: number
  $disabled: boolean
  $hideActiveTrack: boolean
  $labelAlwaysOn: boolean
  $touchNavigation: boolean
}

const RangeWrapper = styled.div.attrs<RangeWrapperProps>(
  ({
    $min,
    $max,
    $valA,
    $valB,
    $disabled,
    $hideActiveTrack,
    $labelAlwaysOn,
    style,
  }) => ({
    'data-disabled': $disabled ? true : null,
    style: {
      '--a': $valA,
      '--b': $valB,
      '--min': $min,
      '--max': $max,
      '--showTooltip': $labelAlwaysOn ? 1 : 0,
      '--background': $disabled
        ? track.entities.indicator.states.disabled.background
        : $hideActiveTrack
          ? 'transparent'
          : track.entities.indicator.background,
      '--background-hover': $hideActiveTrack
        ? 'transparent'
        : track.entities.indicator.states.hover.background,
      ...style,
    },
  }),
)<RangeWrapperProps>`
  --dif: calc(var(--max) - var(--min));
  --realWidth: calc(100% - 12px);
  ${wrapperGrid}
  ${fakeTrackBg}
  &::before,
  &::after {
    ${trackFill};
    background: var(--background);
  }
  /** Faking the active region of the slider */
  &::before {
    margin-left: calc(
      calc(${handle.width} / 2) + (var(--a) - var(--min)) / var(--dif) *
        var(--realWidth)
    );
    width: calc((var(--b) - var(--a)) / var(--dif) * var(--realWidth));
  }

  &::after {
    margin-left: calc(
      calc(${handle.width} / 2) + (var(--b) - var(--min)) / var(--dif) *
        var(--realWidth)
    );
    width: calc((var(--a) - var(--b)) / var(--dif) * var(--realWidth));
  }

  &:has(:focus-visible),
  &:hover {
    & > output {
      --showTooltip: 1;
    }
  }

  @media (hover: hover) and (pointer: fine) {
    &:hover:not([data-disabled]) {
      ${fakeTrackBgHover}
      &::before,
      &::after {
        background: var(--background-hover);
      }
    }
  }

  ${({ $touchNavigation }) =>
    $touchNavigation &&
    css`
      & > input[type='range'] {
        pointer-events: none;
      }

      & > input[type='range']::-webkit-slider-thumb {
        pointer-events: auto;
      }
      & > input[type='range']::-moz-range-thumb {
        pointer-events: auto;
      }
    `};
`

type WrapperProps = {
  $min: number
  $max: number
  $hideActiveTrack: boolean
  $labelAlwaysOn: boolean
} & Pick<SliderProps, 'disabled' | 'value'>

const Wrapper = styled.div.attrs<WrapperProps>(
  ({
    $min,
    $max,
    value,
    disabled,
    $hideActiveTrack,
    $labelAlwaysOn,
    style,
  }) => ({
    style: {
      '--min': $min,
      '--max': $max,
      '--value': value,
      '--showTooltip': $labelAlwaysOn ? 1 : 0,
      '--background': disabled
        ? track.entities.indicator.states.disabled.background
        : $hideActiveTrack
          ? 'transparent'
          : track.entities.indicator.background,
      '--background-hover': $hideActiveTrack
        ? 'transparent'
        : track.entities.indicator.states.hover.background,
      ...style,
    },
  }),
)<WrapperProps>`
  --dif: calc(var(--max) - var(--min));
  --realWidth: calc(100% - 12px);
  ${wrapperGrid}
  ${fakeTrackBg}
  &::after {
    ${trackFill}
    background: var(--background)
  }
  &::after {
    margin-right: calc(
      (var(--max) - var(--value)) / var(--dif) * var(--realWidth)
    );
    /* Adjusting for start dot circle */
    margin-left: 3px;
  }

  &:has(:focus-visible),
  &:hover {
    & > output {
      --showTooltip: 1;
    }
  }

  @media (hover: hover) and (pointer: fine) {
    &:hover:not([disabled]) {
      ${fakeTrackBgHover}
      &::after {
        background: var(--background-hover);
      }
    }
  }
`

const WrapperGroupLabel = styled.div`
  grid-row: 1;
  grid-column: 1 / 3;
`
const WrapperGroupLabelDots = styled(WrapperGroupLabel)`
  &:before,
  &:after {
    content: ' ';
    display: block;
    position: absolute;
    width: ${dot.width};
    height: ${dot.height};
    background: ${tokens.background};
    ${bordersTemplate(dot.border)};
    bottom: ${dot.spacings.bottom};
    left: 0;
  }
  &:after {
    right: 0;
    left: auto;
  }
`

const SrOnlyLabel = styled.label`
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
`

export type SliderProps = {
  /**
   * Id for the elements that labels this slider
   * @deprecated  Use the `aria-labelledby` instead
   * */
  ariaLabelledby?: string
  /** Components value, range of numbers */
  value: number[] | number
  /** Function to be called when value change */
  onChange?: (event: ChangeEvent<HTMLInputElement>, newValue: number[]) => void
  /** Function to be called when value is committed by mouseup event */
  onChangeCommitted?: (
    event: MouseEvent | KeyboardEvent | TouchEvent,
    newValue: number[],
  ) => void
  /** Function for formatting the displayed value. E.g. formatting dates, or adding a unit suffix */
  outputFunction?: (value: number) => string
  /** Max value */
  max?: number
  /**  Min value */
  min?: number
  /** Stepping interval */
  step?: number
  /** Show the min and max dots or not */
  minMaxDots?: boolean
  /** Show the min and max values or not */
  minMaxValues?: boolean
  /** Disabled */
  disabled?: boolean
  /** hides the "active" fill color from the track */
  hideActiveTrack?: boolean
  /** Make the current value label always visible, otherwise it only shows on hover/focus or while using touch input
   * @default false
   */
  labelAlwaysOn?: boolean
} & Omit<HTMLAttributes<HTMLDivElement>, 'onChange'>

export const Slider = forwardRef<HTMLDivElement, SliderProps>(function Slider(
  {
    min = 0,
    max = 100,
    value = [40, 60],
    outputFunction,
    onChange,
    onChangeCommitted,
    minMaxDots = true,
    minMaxValues = true,
    labelAlwaysOn,
    step = 1,
    disabled,
    hideActiveTrack,
    ariaLabelledby,
    'aria-labelledby': ariaLabelledbyNative,
    ...rest
  },
  ref,
) {
  const isNumber = !Array.isArray(value)
  const isRangeSlider = !isNumber && value.length === 2

  const parsedValue: number[] = isNumber ? [value] : value
  const [initalValue, setInitalValue] = useState<number[]>(parsedValue)
  const [sliderValue, setSliderValue] = useState<number[]>(parsedValue)
  const [mousePressed, setMousePressed] = useState<boolean>(false)
  const [touchNavigation, setTouchNavigation] = useState<boolean>(false)

  useEffect(() => {
    if (isRangeSlider) {
      if (value[0] !== initalValue[0] || value[1] !== initalValue[1]) {
        setInitalValue(value)
        setSliderValue(value)
      }
    } else {
      const numberValue = Number(value)
      if (numberValue !== initalValue[0]) {
        setInitalValue([numberValue])
        setSliderValue([numberValue])
      }
    }
  }, [value, initalValue, isRangeSlider])

  const minRange = useRef<HTMLInputElement>(null)
  const maxRange = useRef<HTMLInputElement>(null)
  const onValueChange = (
    event: ChangeEvent<HTMLInputElement>,
    valueArrIdx?: number,
  ) => {
    const changedValue = parseFloat(event.target.value)
    if (isRangeSlider) {
      const newValue = sliderValue.slice()
      newValue[valueArrIdx] = changedValue

      //Prevent min/max values from crossing eachother
      if (valueArrIdx === 0 && newValue[0] >= newValue[1]) {
        newValue[0] = newValue[1]
        maxRange.current?.focus()
      }

      if (valueArrIdx === 1 && newValue[1] <= newValue[0]) {
        newValue[1] = newValue[0]
        minRange.current?.focus()
      }

      setSliderValue(newValue)
      if (onChange) {
        // Callback for provided onChange func
        onChange(event, newValue)
      }
      return
    }

    setSliderValue([changedValue])
    if (onChange) {
      // Callback for provided onChange func
      onChange(event, [changedValue])
    }
  }
  const handleKeyUp = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
      handleCommitedValue(event)
    }
  }

  const handleCommitedValue = (
    event: KeyboardEvent | MouseEvent | TouchEvent,
  ) => {
    if (onChangeCommitted) {
      onChangeCommitted(event, sliderValue)
    }
  }

  const getFormattedText = (text: number) => {
    return outputFunction ? outputFunction(text) : text
  }

  const findClosestRange = (event: MouseEvent | TouchEvent) => {
    const target = event.target as HTMLOutputElement | HTMLInputElement
    if (target.type === 'output' || mousePressed) {
      return
    }
    let clientX: number
    if (event.type === 'touchstart') {
      clientX = (event as TouchEvent<HTMLInputElement>).targetTouches[0].clientX
      setTouchNavigation(true)
    } else if (event.type === 'mousemove') {
      clientX = (event as MouseEvent<HTMLInputElement>).clientX
      setTouchNavigation(false)
    }

    const bounds = target.getBoundingClientRect()
    const x = clientX - bounds.left
    const inputWidth = minRange.current.offsetWidth
    const minValue = parseFloat(minRange.current.value)
    const maxValue = parseFloat(maxRange.current.value)
    const diff = max - min

    const normX = (x / inputWidth) * diff + min

    const maxX = Math.abs(normX - maxValue)
    const minX = Math.abs(normX - minValue)
    if (minX > maxX) {
      minRange.current.style.zIndex = '10'
      maxRange.current.style.zIndex = '20'
    } else {
      minRange.current.style.zIndex = '20'
      maxRange.current.style.zIndex = '10'
    }
    //special cases where both thumbs are all the way to the left or right
    if (minValue === maxValue && minValue === min) {
      minRange.current.style.zIndex = '10'
      maxRange.current.style.zIndex = '20'
    }
    if (minValue === maxValue && maxValue === max) {
      minRange.current.style.zIndex = '20'
      maxRange.current.style.zIndex = '10'
    }
  }

  const handleDragging = (type: string) => {
    if (type === 'mousedown' || type === 'touchmove') {
      setMousePressed(true)
    } else {
      setMousePressed(false)
    }
  }

  let inputIdA = useId(null, 'inputA')
  let inputIdB = useId(null, 'inputB')
  let inputId = useId(null, 'thumb')
  if (rest['id']) {
    const overrideId = rest['id']
    inputIdA = `${overrideId}-thumb-a`
    inputIdB = `${overrideId}-thumb-b`
    inputId = `${overrideId}-thumb`
  }

  const getAriaLabelledby = useCallback(() => {
    if (ariaLabelledbyNative) return ariaLabelledbyNative
    if (ariaLabelledby) return ariaLabelledby
    return null
  }, [ariaLabelledbyNative, ariaLabelledby])

  return (
    <>
      {isRangeSlider ? (
        <RangeWrapper
          {...rest}
          ref={ref}
          role="group"
          aria-labelledby={getAriaLabelledby()}
          $valA={sliderValue[0]}
          $valB={sliderValue[1]}
          $max={max}
          $min={min}
          $disabled={disabled}
          $hideActiveTrack={hideActiveTrack}
          $labelAlwaysOn={labelAlwaysOn || touchNavigation}
          $touchNavigation={touchNavigation}
          onMouseMove={findClosestRange}
          onTouchStartCapture={findClosestRange}
          onTouchEnd={(e) => handleDragging(e.type)}
          onTouchMove={(e) => handleDragging(e.type)}
          onMouseDown={(e) => handleDragging(e.type)}
          onMouseUp={(e) => handleDragging(e.type)}
        >
          {minMaxDots && <WrapperGroupLabelDots />}
          <SrOnlyLabel htmlFor={inputIdA}>Value A</SrOnlyLabel>
          <SliderInput
            type="range"
            ref={minRange}
            value={sliderValue[0]}
            max={max}
            min={min}
            aria-valuemax={max}
            aria-valuemin={min}
            aria-valuenow={sliderValue[0]}
            aria-valuetext={getFormattedText(sliderValue[0]).toString()}
            id={inputIdA}
            step={step}
            onChange={(event) => {
              onValueChange(event, 0)
            }}
            onMouseUp={(event) => handleCommitedValue(event)}
            onTouchEnd={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
            disabled={disabled}
          />
          <Output htmlFor={inputIdA} value={sliderValue[0]}>
            {getFormattedText(sliderValue[0])}
          </Output>
          {minMaxValues && <MinMax>{getFormattedText(min)}</MinMax>}
          <SrOnlyLabel htmlFor={inputIdB}>Value B</SrOnlyLabel>
          <SliderInput
            type="range"
            value={sliderValue[1]}
            min={min}
            max={max}
            aria-valuemax={max}
            aria-valuemin={min}
            aria-valuenow={sliderValue[1]}
            aria-valuetext={getFormattedText(sliderValue[1]).toString()}
            id={inputIdB}
            step={step}
            ref={maxRange}
            onChange={(event) => {
              onValueChange(event, 1)
            }}
            onMouseUp={(event) => handleCommitedValue(event)}
            onTouchEnd={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
            disabled={disabled}
          />
          <Output htmlFor={inputIdB} value={sliderValue[1]}>
            {getFormattedText(sliderValue[1])}
          </Output>
          {minMaxValues && <MinMax>{getFormattedText(max)}</MinMax>}
        </RangeWrapper>
      ) : (
        <Wrapper
          {...rest}
          ref={ref}
          $max={max}
          $min={min}
          value={sliderValue[0]}
          disabled={disabled}
          $hideActiveTrack={hideActiveTrack}
          $labelAlwaysOn={labelAlwaysOn || touchNavigation}
          onTouchStartCapture={() => setTouchNavigation(true)}
          onMouseDownCapture={() => setTouchNavigation(false)}
        >
          <SliderInput
            type="range"
            value={sliderValue[0]}
            min={min}
            max={max}
            aria-valuemax={max}
            aria-valuemin={min}
            aria-valuenow={sliderValue[0]}
            aria-valuetext={getFormattedText(sliderValue[0]).toString()}
            step={step}
            id={inputId}
            onChange={(event) => {
              onValueChange(event)
            }}
            disabled={disabled}
            aria-labelledby={getAriaLabelledby()}
            onMouseUp={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
            onTouchEnd={(event) => handleCommitedValue(event)}
          />
          <Output htmlFor={inputId} value={sliderValue[0]}>
            {getFormattedText(sliderValue[0])}
          </Output>
          {/*  Need an element for pseudo elems :/ */}
          {minMaxDots && <WrapperGroupLabelDots />}
          {minMaxValues && (
            <>
              <MinMax>{getFormattedText(min)}</MinMax>
              <MinMax>{getFormattedText(max)}</MinMax>
            </>
          )}
        </Wrapper>
      )}
    </>
  )
})
