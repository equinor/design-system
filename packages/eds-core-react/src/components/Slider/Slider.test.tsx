/* eslint-disable no-undef */
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { axe } from 'jest-axe'
import { Slider } from './Slider'
import type { SliderProps } from './Slider'

const getUnixTime = (iso: string) => {
  return new Date(iso).getTime()
}

const DateSlider = ({
  value,
  ariaLabelledby = 'date-range-slider',
}: SliderProps) => {
  function outputFunction(val: number) {
    const date = new Date(val)
    // The test node server doesn't have full i18n capabilities, using english is the easiest
    return date.toLocaleDateString('en-US', {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    })
  }

  return (
    <Slider
      min={getUnixTime('2020-01-01')}
      max={getUnixTime('2020-01-31')}
      aria-labelledby={ariaLabelledby}
      step={60 * 60 * 24 * 1000}
      value={value}
      outputFunction={outputFunction}
    />
  )
}

describe('Simple slider', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <Slider value={0} aria-labelledby="test-one" id="test" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(
      <>
        <span id="a11y-test">Text</span>
        <Slider value={40} aria-labelledby="a11y-test" />
      </>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Creates a simple slider when providing a number as value', () => {
    render(<Slider value={0} aria-labelledby="test-one" />)
    const input = screen.getByRole('slider')

    expect(input).toBeDefined()
  })
  it('Sets the aria label', () => {
    render(<Slider value={5} aria-labelledby="test-label" />)
    const input = screen.getByRole('slider')
    expect(input).toHaveAttribute('aria-labelledby', 'test-label')
  })
  it('Can set min, max and step values', () => {
    render(
      <Slider
        value={5}
        min={4}
        max={10}
        step={2}
        aria-labelledby="test-minmax"
      />,
    )
    const input = screen.getByRole('slider')
    expect(input).toHaveAttribute('min', '4')
    expect(input).toHaveAttribute('max', '10')
    expect(input).toHaveAttribute('step', '2')
  })
  it('Updates output according to value change', () => {
    const handleChange = jest.fn()
    render(
      <Slider value={5} aria-labelledby="test-value" onChange={handleChange} />,
    )
    const input = screen.getByRole('slider')
    const outputValue = screen.getByRole('status')
    expect(outputValue).toHaveTextContent('5')
    fireEvent.change(input, { target: { value: '6' } })
    expect(outputValue).toHaveTextContent('6')
    expect(handleChange).toHaveBeenCalledTimes(1)
  })
  it('Can format the output if a formatting function is provided', () => {
    render(
      <DateSlider
        value={getUnixTime('2020-01-01')}
        aria-labelledby="date-test"
      />,
    )
    const input = screen.getByRole('slider')
    const outputValue = screen.getByRole('status')
    expect(outputValue).toHaveTextContent('Wednesday, January 1, 2020')
    expect(outputValue).not.toHaveTextContent('getUnixTime')
    expect(input).toHaveValue(getUnixTime('2020-01-01').toString())
  })
  it('Has minimum and maximum values as default', () => {
    const { container } = render(
      <Slider value={5} aria-labelledby="test-default" min={0} max={100} />,
    )
    expect(container).toHaveTextContent('0')
    expect(container).toHaveTextContent('100')
  })
  it('It is possible to turn off minimum and maximum values', () => {
    const { container } = render(
      <Slider
        value={5}
        aria-labelledby="test-default-off"
        min={0}
        max={100}
        minMaxValues={false}
      />,
    )
    expect(container).not.toHaveTextContent('0')
    expect(container).not.toHaveTextContent('100')
  })
})

describe('Range slider', () => {
  it('Creates a range slider when providing an array as value', () => {
    render(<Slider value={[0, 20]} aria-labelledby="test-array" />)
    const inputs = screen.queryAllByRole('slider')
    expect(inputs).toHaveLength(2)
  })
  it('Can set min, max and step values', () => {
    render(
      <Slider
        value={[5, 6]}
        min={4}
        max={10}
        step={2}
        aria-labelledby="test-range-minmax"
      />,
    )
    const input = screen.getAllByRole('slider')[1]
    expect(input).toHaveAttribute('min', '4')
    expect(input).toHaveAttribute('max', '10')
    expect(input).toHaveAttribute('step', '2')
  })
  it('Updates output according to value change', () => {
    const handleChange = jest.fn()
    const ariaId = 'test-rangechange'
    render(
      <Slider
        value={[3, 6]}
        aria-labelledby={ariaId}
        onChange={handleChange}
      />,
    )

    const inputA = screen.getAllByRole('slider')[0]
    const inputB = screen.getAllByRole('slider')[1]
    const outputA = screen.getAllByRole('status')[0]
    const outputB = screen.getAllByRole('status')[1]

    expect(outputA).toHaveTextContent('3')
    expect(outputB).toHaveTextContent('6')
    fireEvent.change(inputA, { target: { value: '4' } })
    expect(outputA).toHaveTextContent('4')
    expect(handleChange).toHaveBeenCalledTimes(1)
    fireEvent.change(inputB, { target: { value: '7' } })
    expect(outputB).toHaveTextContent('7')
    expect(handleChange).toHaveBeenCalledTimes(2)
  })
  it('Can format the output if a formatting function is provided', () => {
    const ariaId = 'test-output-formatting'
    render(
      <DateSlider
        value={[getUnixTime('2020-01-01'), getUnixTime('2020-01-31')]}
        aria-labelledby={ariaId}
      />,
    )
    const inputA = screen.getAllByRole('slider')[0]
    const inputB = screen.getAllByRole('slider')[1]
    const outputA = screen.getAllByRole('status')[0]
    const outputB = screen.getAllByRole('status')[1]

    expect(outputA).toHaveTextContent('Wednesday, January 1, 2020')
    expect(outputA).not.toHaveTextContent('getUnixTime')
    expect(outputB).toHaveTextContent('Friday, January 31, 2020')
    expect(outputB).not.toHaveTextContent('getUnixTime')
    expect(inputA).toHaveValue(getUnixTime('2020-01-01').toString())
    expect(inputB).toHaveValue(getUnixTime('2020-01-31').toString())
  })

  it('Sets the aria label', () => {
    render(<Slider value={[2, 4]} aria-labelledby="test-rangelabel" />)
    const wrapper = screen.getByRole('group')
    expect(wrapper).toHaveAttribute('aria-labelledby', 'test-rangelabel')
  })
  it('Has minimum and maximum values as default', () => {
    const { container } = render(
      <Slider
        value={[2, 4]}
        aria-labelledby="test-rangedefault"
        min={0}
        max={100}
      />,
    )
    expect(container).toHaveTextContent('0')
    expect(container).toHaveTextContent('100')
  })
  it('It is possible to turn off minimum and maximum values', () => {
    const { container } = render(
      <Slider
        value={[2, 4]}
        aria-labelledby="test-range-minmaxoff"
        min={0}
        max={100}
        minMaxValues={false}
      />,
    )
    expect(container).not.toHaveTextContent('0')
    expect(container).not.toHaveTextContent('100')
  })
})
