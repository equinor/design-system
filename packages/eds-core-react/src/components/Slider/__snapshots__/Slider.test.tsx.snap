// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple slider Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Slider__Wrapper-sc-n1grrg-1 hsFqRl"
    id="test"
    style="--min: 0; --max: 100; --value: 0; --background: var(--eds_interactive_primary__resting, rgba(0, 112, 121, 1));"
    value="0"
  >
    <input
      aria-labelledby="test-one"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      aria-valuetext="0"
      class="SliderInput__StyledSliderInput-sc-17orw4f-0 bYtMJz"
      id="test-thumb"
      max="100"
      min="0"
      step="1"
      type="range"
      value="0"
    />
    <output
      class="Output__StyledOutput-sc-1dy945x-0 iTVTZB"
      for="test-thumb"
      style="--val: 0;"
    >
      0
    </output>
    <div
      class="Slider__WrapperGroupLabel-sc-n1grrg-2 Slider__WrapperGroupLabelDots-sc-n1grrg-3 dxpGeL bWXhHZ"
    />
    <span
      class="MinMax__StyledMinMax-sc-lxmlid-0 eDkejB"
    >
      0
    </span>
    <span
      class="MinMax__StyledMinMax-sc-lxmlid-0 eDkejB"
    >
      100
    </span>
  </div>
</DocumentFragment>
`;
